### PR TITLE
ci: allow running tests that require secrets on PRs from forks

### PR DESCRIPTION
**Pull request by pmalek**
**What this PR does / why we need it**:

In order to allow outside contributions from forks and make tests run (with access to secrets) for those, let's use [`pull_request_target`](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request_target) instead of `pull_request` as an event that would trigger tests workflows.

### CODE REVIEW

**Review by shaneutt — COMMENTED** [review]
.

**Review by shaneutt — CHANGES_REQUESTED** [review]
This would seem to have some security implications. Presently if a threat actor _was_ interested in exfiltrating our CI secrets they would be completely unable to have them READ by their CI runs, even if they managed to get malicious code approved for test runs. Alternatively if we were to make this change, then any PR approved for test runs could exfiltrate the secrets given code crafted for that purpose, correct?

**Comment by rainest** [comment]
I think that risk for licenses is sufficiently mitigated by exfiltration having minimal impact so long as we're not completely oblivious. The test licenses only let you run Kong Enterprise: they cannot be used to escalate to some other malicious capability and do not grant access to any sensitive systems. They expire after a month, so they're not usable long-term and thus not really a piracy concern. If you did have some reason to run Enterprise for a month only, you could just as well get a temporary key by asking sales for a demo environment. Worst case someone with existing approval makes a PR that exfiltrates one, gets a license for one month, and is summarily blocked from interacting with our org and probably raises the ire of GitHub if we report them (or we try to get them charged under the CFAA for extra fun).

Risk for _other_ secrets is a concern. We do not want anyone to exfiltrate our Docker/Pulp/etc. credentials. That would be very bad. We can segregate those in separate environments though, correct? Assuming that's reasonable, my change request would be that we instead audit existing secrets/environments and ensure that credentials are segregated.

AFAIK most of the credential secrets are used in cron workflows, not PR workflows. Are there any credential secrets that we do need for community PRs? Offhand CodeCov is the only one I'm aware of, and it's both disabled here and presumably not the worst thing to get exfiltrated--we'd need to rotate it but IDK what harm you'd do with it (would it let you do anything other than submit a bogus coverage report?).

**Comment by pmalek** [comment]
I do agree with your concerns and I believe there's a middle ground to be met here: require approval from a maintainer every time there's a PR from an outside contributor. This way we'd always need to "OK to test it" (similarly as it's done in kubernetes itself and other CNCF projects, albeit with a bot).

WDYT @shaneutt @rainest ?

**Comment by randmonkey** [comment]
> I think that risk for licenses is sufficiently mitigated by exfiltration having minimal impact so long as we're not completely oblivious. The test licenses only let you run Kong Enterprise: they cannot be used to escalate to some other malicious capability and do not grant access to any sensitive systems. They expire after a month, so they're not usable long-term and thus not really a piracy concern. If you did have some reason to run Enterprise for a month only, you could just as well get a temporary key by asking sales for a demo environment. Worst case someone with existing approval makes a PR that exfiltrates one, gets a license for one month, and is summarily blocked from interacting with our org and probably raises the ire of GitHub if we report them (or we try to get them charged under the CFAA for extra fun).
> 
> Risk for _other_ secrets is a concern. We do not want anyone to exfiltrate our Docker/Pulp/etc. credentials. That would be very bad. We can segregate those in separate environments though, correct? Assuming that's reasonable, my change request would be that we instead audit existing secrets/environments and ensure that credentials are segregated.
> 
> AFAIK most of the credential secrets are used in cron workflows, not PR workflows. Are there any credential secrets that we do need for community PRs? Offhand CodeCov is the only one I'm aware of, and it's both disabled here and presumably not the worst thing to get exfiltrated--we'd need to rotate it but IDK what harm you'd do with it (would it let you do anything other than submit a bogus coverage report?).

We have `secrets.PULP_PASSWORD` in actions `integration-tests-enterprise-postgres` for each PR now. So there might be such risk.

**Comment by shaneutt** [comment]
> I do agree with your concerns and I believe there's a middle ground to be met here: require approval from a maintainer every time there's a PR from an outside contributor. This way we'd always need to "OK to test it" (similarly as it's done in kubernetes itself and other CNCF projects, albeit with a bot).
> 
> WDYT @shaneutt @rainest ?

I don't think that the gate for whether someone can exfiltrate our docker credentials should be as simple as whether someone clicks a button to allow the tests to run. I feel strongly that this is too weak of a barrier. By simply disallowing secret usage for non-maintainers we avoid human error situations on our end

**Comment by shaneutt** [comment]
Sorry the close was unintentional :facepalm:

**Comment by pmalek** [comment]
> I don't think that the gate for whether someone can exfiltrate our docker credentials should be as simple as whether someone clicks a button to allow the tests to run. I feel strongly that this is too weak of a barrier. By simply disallowing secret usage for non-maintainers we avoid human error situations on our end

Hm, that is a valid point. So we have the following options (that I know of);

- Use `pull_request` (what we're doing now), and don't run enterprise tests that require license for PRs from forks. This could initially check only whether the license string is empty (similarly for codecov reports - which are broken anyway now).

- Keep the status quo as of now (use `pull_request`) which will fail enterprise tests for PRs from forks.

- Do something similar what this PR does (use `pull_request_target`) but limit the access to secrets somehow so that we're not as vulnerable.

**Comment by shaneutt** [comment]
> > I don't think that the gate for whether someone can exfiltrate our docker credentials should be as simple as whether someone clicks a button to allow the tests to run. I feel strongly that this is too weak of a barrier. By simply disallowing secret usage for non-maintainers we avoid human error situations on our end
> 
> Hm, that is a valid point. So we have the following options (that I know of);
> 
>     * Use `pull_request` (what we're doing now), and don't run enterprise tests that require license for PRs from forks. This could initially check only whether the license string is empty (similarly for codecov reports - which are broken anyway now).
> 
>     * Keep the status quo as of now (use `pull_request`) which will fail enterprise tests for PRs from forks.
> 
>     * Do something similar what this PR does (use `pull_request_target`) but limit the access to secrets somehow so that we're not as vulnerable.

I'm pretty strongly in favor of option 1, unless we find we have some way to make option 3 work succinctly without the mentioned risks.

**Comment by rainest** [comment]
Yeah, that's reasonable. My point was more that even in the worst-case scenario (we blindly mash the "approve run" button and it executes exfiltration code) the impact isn't all that severe as long as we've isolated the more dangerous secrets in non-test CI environments. I do think we want both for defense in depth--the impact of leaking something like a Docker credential is so severe that I wouldn't want to rely on a single means of protection, especially one that depends on human review--I don't think it's likely we'd actually miss a malicious PR, but humans are fallible, so we shouldn't rely on them alone.

**Comment by pmalek** [comment]
@shaneutt @rainest I do agree with you on the fact that this might be not enough to defend against malicious actors.

> My point was more that even in the worst-case scenario (we blindly mash the "approve run" button and it executes exfiltration code) the impact isn't all that severe as long as we've isolated the more dangerous secrets in non-test CI environments.

I don't think this would. How would you see that implemented @rainest? I mean, if we separate the envs, and let's say put PULP password in one, then regardless of where we've put other secrets this will still be vulnerable to leak if we enable tests workflows that utilize this secret and enable that workflow to run for PRs from forks.

I have not found a way to do it yet but I'd suggest we try:

> Use pull_request (what we're doing now), and don't run enterprise tests that require license for PRs from forks. This could initially check only whether the license string is empty (similarly for codecov reports - which are broken anyway now).

One way or another. Matching PRs authors against a known list of users doesn't scale well and will cost us in maintenance hours down the road. Ideally we could try to check if a user submitting a PR is part of our GH org. I'd say that it'd be safe to run all workflows for those users.

WDYT?

**Comment by rainest** [comment]
> if we separate the envs, and let's say put PULP password in one, then regardless of where we've put other secrets this will still be vulnerable to leak if we enable tests workflows that utilize this secret and enable that workflow to run for PRs from forks.

I'm not sure I follow. Can you provide a sample set of steps that result in a leak? For Pulp at least we'd want to move it to a cron job that updates the license secret. It looks like [that's what we intended and did originally](https://github.com/Kong/team-k8s/issues/27#issuecomment-1143656989), but we switched it back to per test in https://github.com/Kong/go-kong/pull/190 and similar. That mentions updating the secret being an issue, but was there an issue with having the cron job update it?

All said so far, is this actually a concern for `pull_request_target`? The docs for it state:

> This prevents execution of unsafe code from the head of the pull request that could alter your repository or steal any secrets you use in your workflow. This event allows your workflow to do things like label or comment on pull requests from forks. Avoid using this event if you need to build or run code from the pull request.

which makes sense if it is (as I understand to be the case), running the workflow that's currently in the target branch, not any proposed new or updated workflow in the PR. You'd only be able to exfiltrate if we actually merged the change, not merely approve it for testing. However, it also seems like it wouldn't work for tests: my read of the last clause is that it'll also use the _code_ currently in the target and run the tests against that--you can use the target event for things like a bot comment that don't rely on the contents of the PR, but not for things that do.

**Pull request closed without merging**
